File: frontend/lib/services/notification_service.dart
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:timezone/timezone.dart' as tz;
import 'package:timezone/data/latest.dart' as tz_data;
import 'package:barbuddy/utils/constants.dart';
import 'package:barbuddy/models/bac_model.dart';
import 'package:permission_handler/permission_handler.dart';

class NotificationService {
  static final NotificationService _instance = NotificationService._internal();
  
  factory NotificationService() {
    return _instance;
  }
  
  NotificationService._internal();
  
  final FlutterLocalNotificationsPlugin flutterLocalNotificationsPlugin =
      FlutterLocalNotificationsPlugin();
  
  Future<void> init() async {
    tz_data.initializeTimeZones();
    
    const AndroidInitializationSettings initializationSettingsAndroid =
        AndroidInitializationSettings('@mipmap/ic_launcher');
    
    const DarwinInitializationSettings initializationSettingsIOS =
        DarwinInitializationSettings(
      requestSoundPermission: true,
      requestBadgePermission: true,
      requestAlertPermission: true,
    );
    
    const InitializationSettings initializationSettings = InitializationSettings(
      android: initializationSettingsAndroid,
      iOS: initializationSettingsIOS,
    );
    
    await flutterLocalNotificationsPlugin.initialize(
      initializationSettings,
      onDidReceiveNotificationResponse: onNotificationTap,
    );
    
    // Set up notification channels for Android
    await _setupNotificationChannels();
    
    // Request notification permissions
    await requestPermissions();
  }
  
  Future<void> _setupNotificationChannels() async {
    // Safety Alerts Channel
    const AndroidNotificationChannel safetyChannel = AndroidNotificationChannel(
      kSafetyAlertsChannelId,
      'Safety Alerts',
      description: 'Important safety notifications related to your BAC level',
      importance: Importance.high,
    );
    
    // Check-in Reminders Channel
    const AndroidNotificationChannel checkInChannel = AndroidNotificationChannel(
      kCheckInChannelId,
      'Check-in Reminders',
      description: 'Reminders to check in with emergency contacts',
      importance: Importance.high,
    );
    
    // BAC Updates Channel
    const AndroidNotificationChannel bacUpdatesChannel = AndroidNotificationChannel(
      kBACUpdatesChannelId,
      'BAC Updates',
      description: 'Updates about your BAC level and when you\'ll be sober',
      importance: Importance.low,
    );
    
    // Create the channels
    await flutterLocalNotificationsPlugin
        .resolvePlatformSpecificImplementation<
            AndroidFlutterLocalNotificationsPlugin>()
        ?.createNotificationChannels([
      safetyChannel,
      checkInChannel,
      bacUpdatesChannel,
    ]);
  }
  
  Future<void> requestPermissions() async {
    await Permission.notification.request();
  }
  
  void onNotificationTap(NotificationResponse response) {
    // Handle notification taps based on payload
    if (response.payload != null) {
      // Parse payload and handle accordingly
      // e.g. navigate to specific screens
    }
  }
  
  // Show an immediate notification
  Future<void> showNotification({
    required int id,
    required String title,
    required String body,
    String? channelId,
    String? payload,
  }) async {
    final androidDetails = AndroidNotificationDetails(
      channelId ?? kSafetyAlertsChannelId,
      channelId ?? 'Safety Alerts',
      importance: Importance.high,
      priority: Priority.high,
    );
    
    final iOSDetails = const DarwinNotificationDetails(
      presentAlert: true,
      presentBadge: true,
      presentSound: true,
    );
    
    final details = NotificationDetails(
      android: androidDetails,
      iOS: iOSDetails,
    );
    
    await flutterLocalNotificationsPlugin.show(
      id,
      title,
      body,
      details,
      payload: payload,
    );
  }
  
  // Schedule a notification for the future
  Future<void> scheduleNotification({
    required int id,
    required String title,
    required String body,
    required DateTime scheduledTime,
    String? channelId,
    String? payload,
  }) async {
    final androidDetails = AndroidNotificationDetails(
      channelId ?? kSafetyAlertsChannelId,
      channelId ?? 'Safety Alerts',
      importance: Importance.high,
      priority: Priority.high,
    );
    
    final iOSDetails = const DarwinNotificationDetails(
      presentAlert: true,
      presentBadge: true,
      presentSound: true,
    );
    
    final details = NotificationDetails(
      android: androidDetails,
      iOS: iOSDetails,
    );
    
    await flutterLocalNotificationsPlugin.zonedSchedule(
      id,
      title,
      body,
      tz.TZDateTime.from(scheduledTime, tz.local),
      details,
      androidAllowWhileIdle: true,
      uiLocalNotificationDateInterpretation:
          UILocalNotificationDateInterpretation.absoluteTime,
      payload: payload,
    );
  }
  
  // Cancel a specific notification
  Future<void> cancelNotification(int id) async {
    await flutterLocalNotificationsPlugin.cancel(id);
  }
  
  // Cancel all notifications
  Future<void> cancelAllNotifications() async {
    await flutterLocalNotificationsPlugin.cancelAll();
  }
  
  // BAC-specific notifications
  Future<void> scheduleBACNotifications(BACEstimate bacEstimate) async {
    // Cancel any existing BAC notifications
    await cancelNotification(1);
    await cancelNotification(2);
    
    // Only schedule if BAC is above zero
    if (bacEstimate.bac <= 0) {
      return;
    }
    
    // Schedule notification for when BAC reaches legal limit
    if (bacEstimate.bac > kLegalDrivingLimit) {
      await scheduleNotification(
        id: 1,
        title: 'BAC Update',
        body: 'Your BAC is now below the legal driving limit. Remember that impairment can occur at any BAC level.',
        scheduledTime: bacEstimate.legalTime,
        channelId: kBACUpdatesChannelId,
        payload: 'bac_legal',
      );
    }
    
    // Schedule notification for when BAC reaches zero
    await scheduleNotification(
      id: 2,
      title: 'BAC Update',
      body: 'Your estimated BAC has returned to zero.',
      scheduledTime: bacEstimate.soberTime,
      channelId: kBACUpdatesChannelId,
      payload: 'bac_zero',
    );
  }
  
  // Safety alert based on BAC level
  Future<void> showSafetyAlert(BACEstimate bacEstimate) async {
    if (bacEstimate.bac >= kHighBACThreshold) {
      await showNotification(
        id: 3,
        title: 'High BAC Alert',
        body: 'Your BAC is at a high level. DO NOT drive. Please stay hydrated and consider getting assistance if you feel unwell.',
        channelId: kSafetyAlertsChannelId,
        payload: 'high_bac_alert',
      );
    } else if (bacEstimate.bac >= kLegalDrivingLimit) {
      await showNotification(
        id: 4,
        title: 'BAC Above Legal Limit',
        body: 'Your BAC is above the legal driving limit. DO NOT drive. Consider using a ride-sharing service or calling a friend.',
        channelId: kSafetyAlertsChannelId,
        payload: 'legal_limit_alert',
      );
    }
  }
  
  // Schedule check-in reminder
  Future<void> scheduleCheckInReminder(DateTime drinkTime) async {
    final reminderTime = drinkTime.add(Duration(minutes: kCheckInReminderMinutes));
    
    // Only schedule if reminder time is in the future
    if (reminderTime.isAfter(DateTime.now())) {
      await scheduleNotification(
        id: 5,
        title: 'Check-In Reminder',
        body: 'It\'s been a while since your last drink. Would you like to check in with your emergency contacts?',
        scheduledTime: reminderTime,
        channelId: kCheckInChannelId,
        payload: 'check_in_reminder',
      );
    }
  }
  
  // Hydration reminder
  Future<void> scheduleHydrationReminder() async {
    final reminderTime = DateTime.now().add(const Duration(hours: 1));
    
    await scheduleNotification(
      id: 6,
      title: 'Hydration Reminder',
      body: 'Remember to drink water between alcoholic beverages to stay hydrated.',
      scheduledTime: reminderTime,
      channelId: kBACUpdatesChannelId,
      payload: 'hydration_reminder',
    );
  }
}

File: BarBuddy/Shared/Services/NotificationService.swift

import Foundation
import UserNotifications

class NotificationService {
    private let notificationCenter = UNUserNotificationCenter.current()
    
    // Request notification permissions
    func requestPermissions() async -> Bool {
        do {
            let options: UNAuthorizationOptions = [.alert, .sound, .badge]
            return try await notificationCenter.requestAuthorization(options: options)
        } catch {
            print("Error requesting notification permissions: \(error)")
            return false
        }
    }
    
    // Show immediate notification
    func showNotification(
        title: String,
        body: String,
        identifier: String = UUID().uuidString,
        categoryIdentifier: String? = nil
    ) {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default
        
        if let categoryIdentifier = categoryIdentifier {
            content.categoryIdentifier = categoryIdentifier
        }
        
        let request = UNNotificationRequest(
            identifier: identifier,
            content: content,
            trigger: nil // Deliver immediately
        )
        
        notificationCenter.add(request) { error in
            if let error = error {
                print("Error showing notification: \(error)")
            }
        }
    }
    
    // Schedule notification for future time
    func scheduleNotification(
        title: String,
        body: String,
        date: Date,
        identifier: String = UUID().uuidString,
        categoryIdentifier: String? = nil
    ) {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default
        
        if let categoryIdentifier = categoryIdentifier {
            content.categoryIdentifier = categoryIdentifier
        }
        
        let triggerDate = Calendar.current.dateComponents(
            [.year, .month, .day, .hour, .minute, .second],
            from: date
        )
        
        let trigger = UNCalendarNotificationTrigger(
            dateMatching: triggerDate,
            repeats: false
        )
        
        let request = UNNotificationRequest(
            identifier: identifier,
            content: content,
            trigger: trigger
        )
        
        notificationCenter.add(request) { error in
            if let error = error {
                print("Error scheduling notification: \(error)")
            }
        }
    }
    
    // Cancel specific notification
    func cancelNotification(identifier: String) {
        notificationCenter.removePendingNotificationRequests(withIdentifiers: [identifier])
        notificationCenter.removeDeliveredNotifications(withIdentifiers: [identifier])
    }
    
    // Cancel all notifications
    func cancelAllNotifications() {
        notificationCenter.removeAllPendingNotificationRequests()
        notificationCenter.removeAllDeliveredNotifications()
    }
    
    // MARK: - App-specific notifications
    
    // Set up notification categories and actions
    func setUpNotificationCategories() {
        // Emergency check-in category
        let checkInAction = UNNotificationAction(
            identifier: "CHECK_IN_ACTION",
            title: "I'm OK",
            options: .foreground
        )
        
        let callAction = UNNotificationAction(
            identifier: "CALL_ACTION",
            title: "Call Emergency Contact",
            options: .foreground
        )
        
        let emergencyCategory = UNNotificationCategory(
            identifier: "EMERGENCY_CATEGORY",
            actions: [checkInAction, callAction],
            intentIdentifiers: [],
            options: []
        )
        
        // BAC update category
        let viewBACAction = UNNotificationAction(
            identifier: "VIEW_BAC_ACTION",
            title: "View Details",
            options: .foreground
        )
        
        let getRideAction = UNNotificationAction(
            identifier: "GET_RIDE_ACTION",
            title: "Get a Ride",
            options: .foreground
        )
        
        let bacCategory = UNNotificationCategory(
            identifier: "BAC_CATEGORY",
            actions: [viewBACAction, getRideAction],
            intentIdentifiers: [],
            options: []
        )
        
        // Register categories
        notificationCenter.setNotificationCategories([emergencyCategory, bacCategory])
    }
    
    // Schedule BAC update notifications
        func scheduleBACSoberNotification(estimate: BACEstimate) {
            if estimate.bac <= 0 {
                return
            }
            
            // Notification for when BAC reaches legal limit (if applicable)
            if estimate.bac > Constants.BAC.legalLimit {
                scheduleNotification(
                    title: "BAC Update",
                    body: "Your BAC is now below the legal driving limit. Remember that impairment can occur at any BAC level.",
                    date: estimate.legalTime,
                    identifier: "bac_legal_limit",
                    categoryIdentifier: "BAC_CATEGORY"
                )
            }
            
            // Notification for when BAC reaches zero
            scheduleNotification(
                title: "BAC Update",
                body: "Your estimated BAC has returned to zero.",
                date: estimate.soberTime,
                identifier: "bac_zero",
                categoryIdentifier: "BAC_CATEGORY"
            )
        }
        
        // Show safety alert based on BAC level
        func showSafetyAlert(estimate: BACEstimate) {
            if estimate.bac >= Constants.BAC.highThreshold {
                showNotification(
                    title: "High BAC Alert",
                    body: "Your BAC is at a high level. DO NOT drive. Please stay hydrated and consider getting assistance if you feel unwell.",
                    identifier: "high_bac_alert",
                    categoryIdentifier: "BAC_CATEGORY"
                )
            } else if estimate.bac >= Constants.BAC.legalLimit {
                showNotification(
                    title: "BAC Above Legal Limit",
                    body: "Your BAC is above the legal driving limit. DO NOT drive. Consider using a ride-sharing service or calling a friend.",
                    identifier: "legal_limit_alert",
                    categoryIdentifier: "BAC_CATEGORY"
                )
            }
        }
        
        // Schedule check-in reminder
        func scheduleCheckInReminder(drinkTime: Date) {
            let reminderTime = drinkTime.addingTimeInterval(Double(Constants.Time.checkInReminderMinutes * 60))
            
            // Only schedule if reminder time is in the future
            if reminderTime > Date() {
                scheduleNotification(
                    title: "Check-In Reminder",
                    body: "It's been a while since your last drink. Would you like to check in with your emergency contacts?",
                    date: reminderTime,
                    identifier: "check_in_reminder",
                    categoryIdentifier: "EMERGENCY_CATEGORY"
                )
            }
        }
        
        // Hydration reminder
        func scheduleHydrationReminder() {
            let reminderTime = Date().addingTimeInterval(60 * 60) // 1 hour from now
            
            scheduleNotification(
                title: "Hydration Reminder",
                body: "Remember to drink water between alcoholic beverages to stay hydrated.",
                date: reminderTime,
                identifier: "hydration_reminder"
            )
        }
        
        // Schedule emergency check-in if user hasn't interacted with app
        func scheduleInactivityCheckIn(lastActiveTime: Date) {
            let checkInTime = lastActiveTime.addingTimeInterval(2 * 60 * 60) // 2 hours from last activity
            
            // Only schedule if check-in time is in the future
            if checkInTime > Date() {
                scheduleNotification(
                    title: "Are You OK?",
                    body: "We haven't noticed any activity in a while. Tap to check in or alert your emergency contacts.",
                    date: checkInTime,
                    identifier: "inactivity_check_in",
                    categoryIdentifier: "EMERGENCY_CATEGORY"
                )
            }
        }
    }
